IPython/core/magics/packaging.py

```diff
@@ -66,7 +72,11 @@ def pip(self, line):
         Usage:
           %pip install [pkgs]
         """
-        self.shell.system(' '.join([sys.executable, '-m', 'pip', line]))
+        python = sys.executable
+        args = shlex.split(line)
+        command = _format_command(python, "-m", "pip", *args)
```

Comment:
Why this requires `subprocess.list2cmdline`? `list2cmdline` is the windows-style argument splitting, I'm not sure if it is a good idea to mix it with `shlex.split`. Can you explain why a `' '.join` is not sufficient?

Reply:
> Can you explain why a `' '.join` is not sufficient?

Using just `' '.join()` we may lose whitespaces in `sys.executable`.

This command
```sh
$ "/home/my user with spaces/.../python" "-m" "pip" "install 123 456" 
```
may become
```sh
$ "/home/my user with spaces/.../python -m pip install 123 456"
```

So, I guess, we have 2 options here:
- quote whole file path
- escape contained whitespaces

Reply:
I'd go with quoting the file path instead. The responsibility of splitting the command properly based on the OS / platform is on the internals of `shell.system`, so it would be better to leave it there.

Reply:
Updated, please take a look